game_mechanics.py

```diff
@@ -79,23 +81,119 @@ def _make_move(board: np.ndarray, move: Tuple[int, int], current_player: int) ->
-            if has_tile_to_flip(board, move, direction, current_player):
-                return True
+        if has_tile_to_flip(board, move, current_player):
+            return True
     return False
 
 
+def idx_surrounding(arr: np.ndarray, idx: Tuple[int, int]) -> Tuple[np.ndarray, np.ndarray]:
+    """Returns he indexes of the elements within a radius of 1 of the original element (arr must be
+    2d and square currently)"""
+    assert arr.shape[0] == arr.shape[1]
+    dim = arr.shape[0]
+    x, y = idx
+    for num in (x, y):
+        assert 0 <= num < dim
+    x_idx = np.arange(x - 1, x + 2)
+    y_idx = np.arange(y - 1, y + 2)
+
+    # Remove out of array indexes
+    x_idx = x_idx[np.logical_and(x_idx >= 0, x_idx < dim)]
+    y_idx = y_idx[np.logical_and(y_idx >= 0, y_idx < dim)]
+
+    return x_idx, y_idx
+
+
 def is_valid_coord(board_dim: int, row: int, col: int) -> bool:
     return 0 <= row < board_dim and 0 <= col < board_dim
 
 
+def get_vectors(
+    arr: np.ndarray, move: Tuple[int, int]
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """Get all 4 vectors in arr that pass through coordinate move."""
+    dim = arr.shape[0]
+    vertical = arr[np.arange(dim), move[1]]
+    horizontal = arr[move[0], np.arange(dim)]
+
+    # Top left to bottom right diagonal
+    east = dim - max(move)
+    west = min(move)
+    row = np.arange(move[0] - west, move[0] + east)
+    col = np.arange(move[1] - west, move[1] + east)
+    diag1 = arr[row, col]
+
+    # Bottom left to top right diagonal
+    west = min(dim - 1 - move[0], move[1])
+    east = min(move[0], dim - 1 - move[1])
+    row = np.flip(np.arange(move[0] - east, move[0] + west + 1))
+    col = np.arange(move[1] - west, move[1] + east + 1)
+    diag2 = arr[row, col]
+
+    return vertical, horizontal, diag1, diag2
+
+
+@nb.jit(parallel=True)
+def is_sub_arr(a1: np.ndarray, a2: np.ndarray) -> bool:
+    """is a2 a subarray of a1?"""
+    for i in nb.prange(len(a1) - len(a2) + 1):
+        for j in range(len(a2)):
+            if a1[i + j] != a2[j]:
+                break
+        else:
+            return True
+    return False
+
+
+# Current only works for dim<=8 and not optimised for smaller
+MATCHES = [
+    np.array([2, -1, 1]).astype(float),
+    np.array([2, -1, -1, 1]).astype(float),
+    np.array([2, -1, -1, -1, 1]).astype(float),
+    np.array([2, -1, -1, -1, -1, 1]).astype(float),
+    np.array([2, -1, -1, -1, -1, -1, 1]).astype(float),
+    np.array([2, -1, -1, -1, -1, -1, -1, 1]).astype(float),
+    np.array([1, -1, 2]).astype(float),
+    np.array([1, -1, -1, 2]).astype(float),
+    np.array([1, -1, -1, -1, 2]).astype(float),
+    np.array([1, -1, -1, -1, -1, 2]).astype(float),
+    np.array([1, -1, -1, -1, -1, -1, 2]).astype(float),
+    np.array([1, -1, -1, -1, -1, -1, -1, 2]).astype(float),
+]
+
+
+def has_flip(line: np.ndarray) -> bool:
+    for match in MATCHES:
+        if is_sub_arr(line, match):
+            return True
+    return False
```

Comment:
HONKING it into a 1-liner :sunglasses: 

Since it's a generator it stops when it finds one that is true
```suggestion
    return any(is_sub_arr(line, match) for match in MATCHES)
```

Reply:
that is an incredible honk https://i.etsystatic.com/17297128/c/2179/1732/0/316/il/f9dc3d/3848129550/il_340x270.3848129550_1nxl.jpg